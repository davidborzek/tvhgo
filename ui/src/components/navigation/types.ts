import { ReactElement } from 'react';

export type INavigationItem = {
  to: string;
  title: string;
  requiredRoles?: string[];
  icon?: ReactElement;
  items?: INavigationItem[];
};
