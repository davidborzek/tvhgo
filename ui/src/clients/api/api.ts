import axios from 'axios';
import {
  Channel,
  CreateTokenResponse,
  EpgChannel,
  EpgEvent,
  ErrorResponse,
  ListResponse,
  Recording,
  Session,
  Token,
  TwoFactorAuthSettings,
  TwoFactorAuthSetupResult,
  UpdateRecording,
  UpdateUser,
  UpdateUserPassword,
  UserResponse,
} from './api.types';
import qs from 'qs';

type PaginationQuery = {
  limit?: number;
  offset?: number;
};

type SortQuery = {
  sort_key?: string;
  sort_dir?: string;
};

type PaginationSortQuery = PaginationQuery & SortQuery;

export type GetEpgEventsQuery = PaginationSortQuery & {
  title?: string;
  lang?: string;
  channel?: string;
  contentType?: string;
  fullText?: boolean;
  nowPlaying?: boolean;
  durationMin?: number;
  durationMax?: number;
  startsAt?: number;
  endsAt?: number;
};

export type GetEpgChannelEventsQuery = PaginationSortQuery & {
  startsAt?: number;
  endsAt?: number;
};

export type RecordingStatus = 'upcoming' | 'finished' | 'failed' | 'removed';

export type GetRecordingsQuery = PaginationSortQuery & {
  status?: RecordingStatus;
};

export type GetChannelsQuery = PaginationSortQuery & {
  name?: string;
};

export class ApiError extends Error {
  constructor(
    public readonly code: number,
    message: string
  ) {
    super(message);
  }
}

const client = axios.create({
  baseURL: '/api',
  paramsSerializer: (params) => qs.stringify(params, { arrayFormat: 'repeat' }),
});

client.interceptors.response.use(
  (res) => res,
  (err) => {
    if (axios.isAxiosError(err) && err.response) {
      const data = err.response.data as ErrorResponse;
      return Promise.reject(new ApiError(err.response.status, data.message));
    }
    return Promise.reject(err);
  }
);

export async function login(username: string, password: string, code?: string) {
  await client.post('/login', {
    username,
    password,
    totp: code,
  });
}

export async function logout(): Promise<void> {
  await client.post('/logout');
}

export async function getUser(): Promise<UserResponse> {
  const response = await client.get<UserResponse>('/user');
  return response.data;
}

export async function getEpgEvents(
  q?: GetEpgEventsQuery
): Promise<ListResponse<EpgEvent>> {
  const response = await client.get<ListResponse<EpgEvent>>('/epg/events', {
    params: q,
  });
  return response.data;
}

export async function getRelatedEpgEvents(
  id: number,
  q?: PaginationSortQuery
): Promise<ListResponse<EpgEvent>> {
  const response = await client.get<ListResponse<EpgEvent>>(
    `/epg/events/${id}/related`,
    {
      params: q,
    }
  );
  return response.data;
}

export async function getEpgEvent(id: number): Promise<EpgEvent> {
  const response = await client.get<EpgEvent>(`/epg/events/${id}`);
  return response.data;
}

export async function getEpg(
  q?: GetEpgChannelEventsQuery
): Promise<Array<EpgChannel>> {
  const response = await client.get<Array<EpgChannel>>('/epg', {
    params: q,
  });
  return response.data;
}

export async function getChannel(id: string): Promise<Channel> {
  const response = await client.get<Channel>(`/channels/${id}`);
  return response.data;
}

export async function getChannels(
  q?: GetChannelsQuery
): Promise<Array<Channel>> {
  const response = await client.get<Array<Channel>>(`/channels`, {
    params: q,
  });
  return response.data;
}

export async function fetch<T>(path: string): Promise<T> {
  const response = await client.get<T>(path);
  return response.data;
}

export async function recordByEvent(
  eventId: number,
  configId?: string
): Promise<void> {
  await client.post('/recordings/event', {
    eventId,
    configId,
  });
}

export async function stopRecording(id: string): Promise<void> {
  await client.put(`/recordings/${id}/stop`);
}

export async function cancelRecording(id: string): Promise<void> {
  await client.put(`/recordings/${id}/cancel`);
}

export async function removeRecording(id: string): Promise<void> {
  await client.delete(`/recordings/${id}`);
}

export async function stopRecordings(ids: string[]): Promise<void> {
  await client.put(`/recordings/stop`, null, {
    params: { ids },
  });
}

export async function cancelRecordings(ids: string[]): Promise<void> {
  await client.put(`/recordings/cancel`, null, {
    params: { ids },
  });
}

export async function removeRecordings(ids: string[]): Promise<void> {
  await client.delete(`/recordings`, {
    params: { ids },
  });
}

export async function getRecordings(
  q?: GetRecordingsQuery
): Promise<ListResponse<Recording>> {
  const response = await client.get<ListResponse<Recording>>(`/recordings`, {
    params: q,
  });
  return response.data;
}

export async function getRecording(id: string): Promise<Recording> {
  const response = await client.get<Recording>(`/recordings/${id}`);
  return response.data;
}

export async function updateRecording(
  id: string,
  opts: UpdateRecording
): Promise<void> {
  await client.patch(`/recordings/${id}`, opts);
}

export async function updateUser(opts: UpdateUser): Promise<UserResponse> {
  const response = await client.patch<UserResponse>(`/user`, opts);
  return response.data;
}

export async function updateUserPassword(
  opts: UpdateUserPassword
): Promise<void> {
  await client.patch(`/user/password`, opts);
}

export function getRecordingUrl(id: string): string {
  return `/api/recordings/${id}/stream`;
}

export async function getSessions(): Promise<Array<Session>> {
  const response = await client.get<Array<Session>>(`/sessions`);
  return response.data;
}

export async function deleteSession(id: number): Promise<void> {
  return await client.delete(`/sessions/${id}`);
}

export async function getTwoFactorAuthSettings(): Promise<TwoFactorAuthSettings> {
  const response = await client.get<TwoFactorAuthSettings>(`/two-factor-auth`);
  return response.data;
}

export async function deactivateTwoFactorAuth(code: string): Promise<void> {
  await client.put(`/two-factor-auth/deactivate`, {
    code,
  });
}

export async function setupTwoFactorAuth(
  password: string
): Promise<TwoFactorAuthSetupResult> {
  const response = await client.put<TwoFactorAuthSetupResult>(
    `/two-factor-auth/setup`,
    {
      password,
    }
  );
  return response.data;
}

export async function activateTwoFactorAuth(
  password: string,
  code: string
): Promise<void> {
  await client.put(`/two-factor-auth/activate`, {
    password,
    code,
  });
}

export async function getTokens(): Promise<Array<Token>> {
  const response = await client.get<Array<Token>>(`/tokens`);
  return response.data;
}

export async function deleteToken(id: number): Promise<void> {
  await client.delete(`/tokens/${id}`);
}

export async function createToken(name: string): Promise<CreateTokenResponse> {
  const response = await client.post<CreateTokenResponse>(`/tokens`, { name });
  return response.data;
}
