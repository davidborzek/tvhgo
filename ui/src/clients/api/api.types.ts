export type ListResponse<T> = {
  entries: T[];
  total: number;
  offset: number;
};

export type UserResponse = {
  id: string;
  username: string;
  email: string;
  displayNAme: string;
  createdAt: number;
  updatedAt: number;
};

export type ErrorResponse = {
  message: string;
};

export type EpgEvent = {
  id: number;
  audioDesc: boolean;
  channelId: string;
  channelName: string;
  channelNumber: number;
  piconId: number;
  description: string;
  endsAt: number;
  hd: boolean;
  nextEventId: number;
  startsAt: number;
  subtitle: string;
  subtitled: boolean;
  title: string;
  widescreen: boolean;
  dvrUuid?: string;
  dvrState?: string;
};

export type EpgChannel = {
  channelId: string;
  channelName: string;
  channelNumber: number;
  piconId: number;
  events: EpgEvent[];
};

export type Channel = {
  id: string;
  enabled: boolean;
  name: string;
  number: number;
  piconId: number;
};
