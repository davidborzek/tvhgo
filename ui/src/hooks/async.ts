import { useEffect } from 'react';

type PromiseFun = () => Promise<unknown>;

export const usePromiseAll = (values: PromiseFun[]) => {
  useEffect(() => {
    Promise.all(values.map((fn) => fn()));
  }, []);
};
