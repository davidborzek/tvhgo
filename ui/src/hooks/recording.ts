import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
  ApiError,
  cancelRecording,
  cancelRecordings,
  createRecording,
  getRecording,
  getRecordings,
  GetRecordingsQuery,
  recordByEvent,
  removeRecording,
  removeRecordings,
  stopRecording,
  stopRecordings,
  updateRecording,
} from '@/clients/api/api';
import {
  CreateRecordingOpts,
  Recording,
  UpdateRecording,
} from '@/clients/api/api.types';
import { useLoading } from '@/contexts/LoadingContext';
import { useNotification } from './notification';

export const useManageRecordingByEvent = () => {
  const { notifyError, notifySuccess, dismissNotification } = useNotification(
    'manageRecordingByEvent'
  );

  const { t } = useTranslation();
  const [pending, setPending] = useState(false);

  const createRecording = async (eventId: number) => {
    dismissNotification();
    setPending(true);
    return await recordByEvent(eventId)
      .then(() => {
        notifySuccess(t('recording_created'));
      })
      .catch(() => {
        notifyError(t('unexpected'));
      })
      .finally(() => {
        setPending(false);
      });
  };

  const _cancelRecording = async (id: string, success?: () => void) => {
    dismissNotification();
    setPending(true);
    return await cancelRecording(id)
      .then(() => {
        notifySuccess(t('recording_canceled'));
        success && success();
      })
      .catch(() => {
        notifyError(t('unexpected'));
      })
      .finally(() => {
        setPending(false);
      });
  };

  const _stopRecording = async (id: string, success?: () => void) => {
    dismissNotification();
    setPending(true);
    return await stopRecording(id)
      .then(() => {
        notifySuccess(t('recording_stopped'));
        success && success();
      })
      .catch(() => {
        notifyError(t('unexpected'));
      })
      .finally(() => {
        setPending(false);
      });
  };

  const _removeRecording = async (id: string, success?: () => void) => {
    dismissNotification();
    setPending(true);
    return await removeRecording(id)
      .then(() => {
        notifySuccess(t('recording_removed'));
        success && success();
      })
      .catch(() => {
        notifyError(t('unexpected'));
      })
      .finally(() => {
        setPending(false);
      });
  };

  const _updateRecording = async (id: string, opts: UpdateRecording) => {
    dismissNotification();
    setPending(true);
    return await updateRecording(id, opts)
      .then(() => {
        notifySuccess(t('recording_updated'));
      })
      .catch(() => {
        notifyError(t('unexpected'));
      })
      .finally(() => {
        setPending(false);
      });
  };

  return {
    createRecording,
    cancelRecording: _cancelRecording,
    stopRecording: _stopRecording,
    removeRecording: _removeRecording,
    updateRecording: _updateRecording,
    pending,
  };
};

export const useFetchRecordings = (q?: GetRecordingsQuery) => {
  const { t } = useTranslation();

  const { setIsLoading } = useLoading();

  const [error, setError] = useState<string | null>(null);
  const [recordings, setRecordings] = useState<Recording[] | null>(null);
  const [total, setTotal] = useState(0);

  const fetch = () => {
    setIsLoading(true);
    setRecordings([]);
    getRecordings(q)
      .then((result) => {
        setRecordings(result.entries);
        setTotal(result.total);
      })
      .catch(() => setError(t('unexpected')))
      .finally(() => setIsLoading(false));
  };

  useEffect(() => {
    fetch();
  }, [q?.status, q?.limit, q?.offset]);

  return { recordings, error, fetch, total };
};

export const useFetchRecording = () => {
  const { t } = useTranslation();
  const { setIsLoading } = useLoading();

  const [error, setError] = useState<string | null>(null);
  const [recording, setRecording] = useState<Recording>();

  const fetch = async (id: string) => {
    setIsLoading(true);
    getRecording(id)
      .then(setRecording)
      .catch((error) => {
        if (error instanceof ApiError && error.code === 404) {
          setError(t('not_found'));
        } else {
          setError(t('unexpected'));
        }
      })
      .finally(() => setIsLoading(false));
  };

  return { error, recording, fetch };
};

export const useManageRecordings = () => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('manageRecordings');

  const { t } = useTranslation();
  const [pending, setPending] = useState(false);

  const stopAndCancelRecordings = async (
    stopIds: string[],
    cancelIds: string[]
  ) => {
    dismissNotification();
    setPending(true);

    const promises = [];
    if (stopIds.length > 0) {
      promises.push(stopRecordings(stopIds));
    }

    if (cancelIds.length > 0) {
      promises.push(cancelRecordings(cancelIds));
    }

    return await Promise.all(promises)
      .then(() => {
        notifySuccess(t('recordings_stopped_canceled'));
      })
      .catch(() => {
        notifyError(t('unexpected'));
      })
      .finally(() => {
        setPending(false);
      });
  };

  const _removeRecordings = async (ids: string[]) => {
    dismissNotification();
    setPending(true);

    return await removeRecordings(ids)
      .then(() => {
        notifySuccess(t('recordings_removed'));
      })
      .catch(() => {
        notifyError(t('unexpected'));
      })
      .finally(() => {
        setPending(false);
      });
  };

  return {
    stopAndCancelRecordings,
    removeRecordings: _removeRecordings,
    pending,
  };
};

export const useCreateRecording = () => {
  const { notifyError, notifySuccess, dismissNotification } =
    useNotification('createRecording');

  const { t } = useTranslation();
  const [pending, setPending] = useState(false);

  const _createRecording = async (opts: CreateRecordingOpts) => {
    dismissNotification();
    setPending(true);

    createRecording(opts)
      .catch((error) => {
        notifyError(t('unexpected'));
      })
      .finally(() => {
        setPending(false);
      });
  };

  return { createRecording: _createRecording, pending };
};
