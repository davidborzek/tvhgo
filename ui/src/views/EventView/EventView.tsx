import { useEffect } from 'react';
import { useNavigate, useParams } from 'react-router-dom';

import Error from '@/components/Error/Error';
import EventChannelInfo from '@/components/Event/EventChannelInfo/EventChannelInfo';
import EventInfo from '@/components/Event/EventInfo/EventInfo';
import EventRelated from '@/components/Event/EventRelated/EventRelated';
import { useFetchEvent } from '@/hooks/epg';
import { useManageRecordingByEvent } from '@/hooks/recording';

import styles from './EventView.module.scss';

function EventView() {
  const navigate = useNavigate();
  const params = useParams();
  const { fetch, error, event, relatedEvents } = useFetchEvent();
  const { createRecording, pending } = useManageRecordingByEvent();

  const fetchEvent = () => {
    const id = params['id'];
    if (id) {
      fetch(parseInt(id));
    }
  };

  useEffect(() => {
    fetchEvent();
  }, [params]);

  const handleOnRecord = async () => {
    if (!event) {
      return;
    }

    if (event.dvrUuid) {
      navigate(`/recordings/${event.dvrUuid}`);
      return;
    }

    await createRecording(event.id);
    fetchEvent();
  };

  if (error) {
    return <Error message={error} />;
  }

  if (!event) {
    return <></>;
  }

  return (
    <div className={styles.Event}>
      <EventChannelInfo
        channelName={event.channelName}
        picon={`/api/picon/${event.piconId}`}
      />
      <EventInfo
        event={event}
        handleOnRecord={handleOnRecord}
        pending={pending}
      />
      <EventRelated relatedEvents={relatedEvents} />
    </div>
  );
}

export default EventView;
