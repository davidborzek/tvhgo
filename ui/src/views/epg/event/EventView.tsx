import {
  LoaderFunctionArgs,
  useLoaderData,
  useNavigate,
  useRevalidator,
} from 'react-router-dom';

import EventChannelInfo from '@/components/epg/event/channelInfo/EventChannelInfo';
import EventInfo from '@/components/epg/event/info/EventInfo';
import EventRelated from '@/components/epg/event/related/EventRelated';
import { useManageRecordingByEvent } from '@/hooks/recording';

import styles from './EventView.module.scss';
import { getEpgEvent, getRelatedEpgEvents } from '@/clients/api/api';
import { EpgEvent } from '@/clients/api/api.types';

export async function loader({ params }: LoaderFunctionArgs) {
  if (!params.id) {
    return;
  }

  const id = parseInt(params.id);
  if (!id) {
    return;
  }

  const [event, related] = await Promise.all([
    getEpgEvent(id),
    getRelatedEpgEvents(id),
  ]);

  return [event, related.entries.filter((r) => r.id !== id)];
}

export function Component() {
  const navigate = useNavigate();
  const revalidator = useRevalidator();
  const { createRecording, pending } = useManageRecordingByEvent();
  const [event, relatedEvents] = useLoaderData() as [EpgEvent, Array<EpgEvent>];

  const handleOnRecord = async () => {
    if (!event) {
      return;
    }

    if (event.dvrUuid) {
      navigate(`/recordings/${event.dvrUuid}`);
      return;
    }

    await createRecording(event.id);
    revalidator.revalidate();
  };

  return (
    <div className={styles.Event}>
      <EventChannelInfo
        channelName={event.channelName}
        picon={`/api/picon/${event.piconId}`}
      />
      <EventInfo
        event={event}
        handleOnRecord={handleOnRecord}
        pending={pending}
      />
      <EventRelated relatedEvents={relatedEvents} />
    </div>
  );
}

Component.displayName = 'EventView';
