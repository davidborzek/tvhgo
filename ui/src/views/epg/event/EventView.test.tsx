import { DVRConfig, EpgEvent } from '@/clients/api/api.types';
import { afterEach, beforeEach, expect, test, vi } from 'vitest';
import { cleanup, render } from '@testing-library/react';
import { useLoaderData, useNavigate, useRevalidator } from 'react-router-dom';

import { Component as EventView } from './EventView';
import { useManageRecordingByEvent } from '@/hooks/recording';
import userEvent from '@testing-library/user-event';

vi.mock('@/hooks/epg');
vi.mock('@/hooks/recording');
vi.mock('react-router-dom');

const EVENT_ID = 1;
const RECORDING_ID = 'someRecordingID';

const buildEvent = (dvrUuid?: string): EpgEvent => {
  return {
    audioDesc: true,
    channelId: 'someChannelId',
    channelName: 'Some Channel',
    channelNumber: 1,
    description: 'Some Description',
    endsAt: 0,
    hd: true,
    id: EVENT_ID,
    nextEventId: 2,
    piconId: 1,
    startsAt: 0,
    subtitle: 'Some Subtitle',
    subtitled: true,
    title: 'Some Title',
    widescreen: true,
    dvrUuid,
  };
};

const relatedEvents: EpgEvent[] = [
  {
    audioDesc: true,
    channelId: 'someChannelId',
    channelName: 'Some Channel',
    channelNumber: 1,
    description: 'Some Description 2',
    endsAt: 0,
    hd: true,
    id: 2,
    nextEventId: 3,
    piconId: 1,
    startsAt: 0,
    subtitle: 'Some Subtitle 2',
    subtitled: true,
    title: 'Some Title 2',
    widescreen: true,
  },
  {
    audioDesc: true,
    channelId: 'someChannelId2',
    channelName: 'Some Channel2',
    channelNumber: 1,
    description: 'Some Description 3',
    endsAt: 0,
    hd: true,
    id: 3,
    nextEventId: 4,
    piconId: 1,
    startsAt: 0,
    subtitle: 'Some Subtitle 3',
    subtitled: true,
    title: 'Some Title 3',
    widescreen: true,
  },
];

const buildDvrProfile = (id: string): DVRConfig => {
  return {
    id,
    artwork: {
      allowUnidentifiableBroadcasts: false,
      commandLineOptions: '',
      fetch: true,
    },
    clone: false,
    deleteAfterPlayback: 0,
    enabled: true,
    endPadding: 2,
    epg: {
      autorec: {
        maxCount: 0,
        maxSchedules: 0,
      },
      duplicateHandling: 'foobar',
      epgRunning: true,
      epgUpdateWindow: 0,
      skipCommercials: true,
    },
    file: {
      allowWhitespace: true,
      cleanTitle: true,
      includeChannel: true,
      includeDate: true,
      includeEpisode: true,
      includeSubtitle: true,
      includeTime: true,
      omitTitle: true,
      tagFiles: true,
      windowsCompatibleFilename: false,
    },
    hooks: {
      remove: '',
      start: '',
      stop: '',
    },
    name: 'Some Profile',
    original: false,
    priority: 'high',
    recordingFileRetention: {
      days: 0,
      type: 'forever',
    },
    recordingInfoRetention: {
      days: 0,
      type: 'forever',
    },
    rerecordErrors: 0,
    startPadding: 0,
    storage: {
      cacheScheme: 'none',
      charset: 'utf-8',
      directoryPermissions: '0755',
      filePermissions: '0644',
      maintainFreeSpace: 0,
      maintainUsedSpace: 0,
      pathnameFormat: 'foo$q$bar',
      path: '/foo/bar',
    },
    streamProfileId: 'someStreamProfileId',
    subdirectories: {
      channelSubdir: true,
      daySubdir: true,
      titleSubdir: true,
      tvMoviesSubdirFormat: '',
      tvShowsSubdirFormat: '',
    },
    tunerWarmUpTime: 0,
  };
};

const navigateMock = vi.fn();
const revalidateMock = vi.fn();

beforeEach(() => {
  vi.mocked(useNavigate).mockReturnValue(navigateMock);
  vi.mocked(useRevalidator).mockReturnValue({
    revalidate: revalidateMock,
    state: 'idle',
  });
});

afterEach(() => {
  vi.resetAllMocks();
  cleanup();
});

test('should render without related events', () => {
  vi.mocked(useLoaderData).mockReturnValue([
    buildEvent(),
    [],
    [buildDvrProfile('someProfileId')],
  ]);

  vi.mocked(useManageRecordingByEvent).mockReturnValue({
    pending: false,
    createRecording: async () => {},
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any);

  const document = render(<EventView />);

  expect(document.asFragment()).toMatchSnapshot();
});

test('should render with recording', () => {
  vi.mocked(useLoaderData).mockReturnValue([
    buildEvent(RECORDING_ID),
    [],
    [buildDvrProfile('someProfileId')],
  ]);

  vi.mocked(useManageRecordingByEvent).mockReturnValue({
    pending: false,
    createRecording: async () => {},
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any);

  const document = render(<EventView />);

  expect(document.asFragment()).toMatchSnapshot();
});

test('should render with pending button', () => {
  vi.mocked(useLoaderData).mockReturnValue([
    buildEvent(),
    [],
    [buildDvrProfile('someProfileId')],
  ]);

  vi.mocked(useManageRecordingByEvent).mockReturnValue({
    pending: true,
    createRecording: async () => {},
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any);

  const document = render(<EventView />);

  expect(document.asFragment()).toMatchSnapshot();
});

test('should render with related events', () => {
  vi.mocked(useLoaderData).mockReturnValue([
    buildEvent(),
    relatedEvents,
    [buildDvrProfile('someProfileId')],
  ]);

  vi.mocked(useManageRecordingByEvent).mockReturnValue({
    pending: false,
    createRecording: async () => {},
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any);

  const document = render(<EventView />);

  expect(document.asFragment()).toMatchSnapshot();
});

test('should create recording', async () => {
  vi.mocked(useLoaderData).mockReturnValue([
    buildEvent(),
    [],
    [buildDvrProfile('someProfileId')],
  ]);

  const createRecordingMock = vi.fn();
  vi.mocked(useManageRecordingByEvent).mockReturnValue({
    pending: false,
    createRecording: createRecordingMock,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any);

  const document = render(<EventView />);

  await userEvent.click(document.getByText('record'));
  expect(createRecordingMock).toHaveBeenNthCalledWith(1, EVENT_ID, undefined);
});

test('should open profile selection', async () => {
  vi.mocked(useLoaderData).mockReturnValue([
    buildEvent(),
    [],
    [buildDvrProfile('someProfileId'), buildDvrProfile('someProfileId')],
  ]);

  const createRecordingMock = vi.fn();
  vi.mocked(useManageRecordingByEvent).mockReturnValue({
    pending: false,
    createRecording: createRecordingMock,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any);

  const document = render(<EventView />);

  await userEvent.click(document.getByText('record'));
  expect(document.asFragment()).toMatchSnapshot();
});

test('should navigate to recording', async () => {
  vi.mocked(useLoaderData).mockReturnValue([
    buildEvent(RECORDING_ID),
    [],
    [buildDvrProfile('someProfileId')],
  ]);

  vi.mocked(useManageRecordingByEvent).mockReturnValue({
    pending: false,
    createRecording: async () => {},
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any);

  const document = render(<EventView />);
  await userEvent.click(document.getByText('modify_recording'));

  expect(navigateMock).toHaveBeenCalledWith(`/dvr/recordings/${RECORDING_ID}`);
});
