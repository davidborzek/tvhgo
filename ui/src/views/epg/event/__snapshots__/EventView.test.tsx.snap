// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render with pending button 1`] = `
<DocumentFragment>
  <div
    class="_Event_6a4861"
  >
    <div
      class="_EventChannelInfo_e37c34"
    >
      <span
        class="_channelName_e37c34"
      >
        Some Channel
      </span>
      <div
        class="_picon_e37c34"
      >
        <div
          style="width: 100%; height: 100%;"
        />
        <img
          class="_picon_e37c34"
          src="/api/picon/1"
          style="display: none;"
          title="Some Channel"
        />
      </div>
    </div>
    <div
      class="_EventInfo_a6cff5"
    >
      <h1>
        Some Title
      </h1>
      <div>
        <button
          class="_EventRecordButton_1a746f _pending_1a746f"
        >
          <svg
            class="_icon_1a746f"
            fill="#000000"
            height="800px"
            id="Layer_1"
            viewBox="0 0 485 485"
            width="800px"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <g>
              <path
                d="M0,0v485h485V0H0z M455,455H30V30h425V455z"
              />
              <circle
                cx="242.5"
                cy="242.5"
                r="102.5"
              />
            </g>
          </svg>
          record
        </button>
      </div>
      <div
        class="_PairList_768617"
      >
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            subtitle
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            Some Subtitle
          </div>
        </div>
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            airs
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            event_datetime
          </div>
        </div>
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            description
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            Some Description
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render with recording 1`] = `
<DocumentFragment>
  <div
    class="_Event_6a4861"
  >
    <div
      class="_EventChannelInfo_e37c34"
    >
      <span
        class="_channelName_e37c34"
      >
        Some Channel
      </span>
      <div
        class="_picon_e37c34"
      >
        <div
          style="width: 100%; height: 100%;"
        />
        <img
          class="_picon_e37c34"
          src="/api/picon/1"
          style="display: none;"
          title="Some Channel"
        />
      </div>
    </div>
    <div
      class="_EventInfo_a6cff5"
    >
      <h1>
        Some Title
      </h1>
      <div>
        <button
          class="_EventRecordButton_1a746f"
        >
          <svg
            class="_icon_1a746f"
            fill="#000000"
            height="800px"
            id="Layer_1"
            viewBox="0 0 485 485"
            width="800px"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <g>
              <path
                d="M0,0v485h485V0H0z M455,455H30V30h425V455z"
              />
              <circle
                cx="242.5"
                cy="242.5"
                r="102.5"
              />
            </g>
          </svg>
          modify_recording
        </button>
      </div>
      <div
        class="_PairList_768617"
      >
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            subtitle
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            Some Subtitle
          </div>
        </div>
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            airs
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            event_datetime
          </div>
        </div>
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            description
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            Some Description
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render with related events 1`] = `
<DocumentFragment>
  <div
    class="_Event_6a4861"
  >
    <div
      class="_EventChannelInfo_e37c34"
    >
      <span
        class="_channelName_e37c34"
      >
        Some Channel
      </span>
      <div
        class="_picon_e37c34"
      >
        <div
          style="width: 100%; height: 100%;"
        />
        <img
          class="_picon_e37c34"
          src="/api/picon/1"
          style="display: none;"
          title="Some Channel"
        />
      </div>
    </div>
    <div
      class="_EventInfo_a6cff5"
    >
      <h1>
        Some Title
      </h1>
      <div>
        <button
          class="_EventRecordButton_1a746f"
        >
          <svg
            class="_icon_1a746f"
            fill="#000000"
            height="800px"
            id="Layer_1"
            viewBox="0 0 485 485"
            width="800px"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <g>
              <path
                d="M0,0v485h485V0H0z M455,455H30V30h425V455z"
              />
              <circle
                cx="242.5"
                cy="242.5"
                r="102.5"
              />
            </g>
          </svg>
          record
        </button>
      </div>
      <div
        class="_PairList_768617"
      >
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            subtitle
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            Some Subtitle
          </div>
        </div>
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            airs
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            event_datetime
          </div>
        </div>
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            description
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            Some Description
          </div>
        </div>
      </div>
    </div>
    <div
      class="_EventRelated_9241f8"
    >
      <h2>
        related_events
      </h2>
      <div
        class="_PairList_768617"
      >
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            Some Channel
          </div>
          <div
            class="_PairValue_05cc8e"
          />
        </div>
      </div>
      <div
        class="_PairList_768617"
      >
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            Some Channel2
          </div>
          <div
            class="_PairValue_05cc8e"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render without related events 1`] = `
<DocumentFragment>
  <div
    class="_Event_6a4861"
  >
    <div
      class="_EventChannelInfo_e37c34"
    >
      <span
        class="_channelName_e37c34"
      >
        Some Channel
      </span>
      <div
        class="_picon_e37c34"
      >
        <div
          style="width: 100%; height: 100%;"
        />
        <img
          class="_picon_e37c34"
          src="/api/picon/1"
          style="display: none;"
          title="Some Channel"
        />
      </div>
    </div>
    <div
      class="_EventInfo_a6cff5"
    >
      <h1>
        Some Title
      </h1>
      <div>
        <button
          class="_EventRecordButton_1a746f"
        >
          <svg
            class="_icon_1a746f"
            fill="#000000"
            height="800px"
            id="Layer_1"
            viewBox="0 0 485 485"
            width="800px"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <g>
              <path
                d="M0,0v485h485V0H0z M455,455H30V30h425V455z"
              />
              <circle
                cx="242.5"
                cy="242.5"
                r="102.5"
              />
            </g>
          </svg>
          record
        </button>
      </div>
      <div
        class="_PairList_768617"
      >
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            subtitle
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            Some Subtitle
          </div>
        </div>
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            airs
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            event_datetime
          </div>
        </div>
        <div
          class="_Pair_2d5b35"
        >
          <div
            class="_PairKey_b66d6a"
          >
            description
          </div>
          <div
            class="_PairValue_05cc8e"
          >
            Some Description
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
